### PR TITLE
elastic net

### PR DESCRIPTION
**Pull request by rayeaster**
https://github.com/haifengl/smile/issues/351

### CODE REVIEW

**Comment by rayeaster** [comment]
This implementation allows reduced to pure LASSO (L1) but not allowed to reduced to pure RIDGE (L2) since it use a modified version of LASSO for existing solver in smile

**Comment by haifengl** [comment]
ElasticNet is essentially equivalent linear SVM. Have you compared the accuracy between them? Thanks.

**Comment by rayeaster** [comment]
Sure. Will do the comparison between ElasticNet and Linear SVM.

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+ * 
+ * @author rayeaster
+ */
+public class ElasticNet implements Regression<double[]>, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * hyperparameter for regularization
+     */
+    private double lambda = 0.1;
```
[the remark this answers is not in the document]
Comment (rayeaster) [reply core/src/main/java/smile/regression/ElasticNet.java]: lamdba is the absolute regularization parameter while lamdba1p and lamdba2p are portions for L1 and L2, with their sum equals to 1

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                    ret[i][j] = c() * x[i][j];
+                }
+            } else {
+                ret[i][i - x.length] = c() * Math.sqrt(lambda2);
+            }
+        }
+        return ret;
+    }
+
+    private Matrix getModifiedData(Matrix x) {
```
[the remark this answers is not in the document]
Comment (rayeaster) [reply core/src/main/java/smile/regression/ElasticNet.java]: This is consistent with the formula I referenced to reduce ElasticeNet to a LASSO problem.

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        this.lambda1p = lambda1p;
+        this.lambda2p = 1 - lambda1p;
+        this.lambda1 = lambda1p * lambda;
+        this.lambda2 = lambda2p * lambda;
+        this.p = x.ncols();
+        modifiedLasso = new LASSO(getModifiedData(x), getModifiedResponse(y), getModifiedLambda(), tol, maxIter);
+
+        correctedW = new double[modifiedLasso.coefficients().length];
+        for (int i = 0; i < correctedW.length; i++) {
+            correctedW[i] = c() * modifiedLasso.coefficients()[i];
```
[the remark this answers is not in the document]
Comment (rayeaster) [reply core/src/main/java/smile/regression/ElasticNet.java]: I use the so-called "vanilla" solution. But yes, your suggested is a better estimate according to MLAPP book section 13.5.3.3

File: core/src/test/java/smile/regression/SVRTest.java [hunk 4]
```diff
@@ -16,9 +16,13 @@
 
 package smile.regression;
 
+import smile.math.kernel.LinearKernel;
```
[the remark this answers is not in the document]
Comment (rayeaster) [reply core/src/test/java/smile/regression/SVRTest.java]: > No need to modify SVM test then.

you mean no need to modify SVR test, right?

[the remark this answers is not in the document]
Comment (rayeaster) [reply core/src/test/java/smile/regression/SVRTest.java]: Remove modification to SVR test and add test on a binary SVM too

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    private double lambda2 = 0.1;
+    /**
+     * The dimensionality.
+     */
+    private int p;
+    /**
+     * corrected coefficients
+     */
+    private double[] w = null;
```
Comment (haifengl) [inline core/src/main/java/smile/regression/ElasticNet.java]: no need of initialization. JVM does it.

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    /**
+     * corrected intercept
+     */
+    private double b;
+
+    /** reduced to lasso problem */
+    private LASSO lasso = null;
```
Comment (haifengl) [inline core/src/main/java/smile/regression/ElasticNet.java]: same as above

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    public double predict(double[] x) {
+        if (x.length != p) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid input vector size: %d, expected: %d", x.length, p));
+        }
+
+        return Math.dot(x, w) + b;
+    }
+
+    public double[] coefficients() {
```
Comment (haifengl) [inline core/src/main/java/smile/regression/ElasticNet.java]: All public methods should have documents.

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        return w;
+    }
+
+    public LASSO lasso() {
```
Comment (haifengl) [inline core/src/main/java/smile/regression/ElasticNet.java]: see above

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+     * 
+     * @param x
+     *            the original data array
+     * @return data with padding
+     */
+    private double[][] getAugmentedData(double[][] x) {
+        double[][] ret = new double[x.length + p][p];
+        double padding = c * Math.sqrt(lambda2);
+        for (int i = 0; i < ret.length; i++) {
+            if (i <= x.length - 1) {
```
Comment (haifengl) [inline core/src/main/java/smile/regression/ElasticNet.java]: It is faster if we split this loop over into two. for (i = 0; i < x.length) and (i=x.length; i < ret.length).

File: core/src/test/java/smile/classification/SVMTest.java [hunk 10]
```diff
@@ -30,6 +30,8 @@
 import smile.math.kernel.GaussianKernel;
 import smile.math.kernel.LinearKernel;
 import smile.math.kernel.PolynomialKernel;
+import smile.regression.ElasticNet;
```
Comment (haifengl) [inline core/src/test/java/smile/classification/SVMTest.java]: The equivalent SVM of ElasticNet is not this straightforward. Let's skip this for now.

File: core/src/main/java/smile/regression/ElasticNet.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        return Math.dot(x, w) + b;
+    }
+
+    /**
+     * @return the linear coefficients.
+     */
+    public double[] coefficients() {
+        return w;
+    }
+
```
[the remark this answers is not in the document]
Comment (rayeaster) [reply core/src/main/java/smile/regression/ElasticNet.java]: > Please also add
> 
> ```
>     /**
>      * Returns the intercept.
>      */
>     public double intercept() {
>         return b;
>     }
> ```

add what?

**Comment by haifengl** [comment]
Thanks a lot!

**Pull request merged**